test/ec_comp_test.c

```diff
@@ -2066,235 +86,369 @@ static int init_curves(void)
+            return 0;
+    } else {
+        /* If a parameter format was not specified, expect named curve */
+        if (!TEST_int_eq(OPENSSL_EC_NAMED_CURVE, asn1_flag))
+            return 0;
+    }
+
+    return 1;
 }
 
-#ifndef GENERATE_KNOWN_RESULTS
-static int verify_known_result(size_t curve_i, size_t param_format_i, size_t comp_format_i, size_t key_src, struct default_config_st *dcfg)
+static int test_reserialize(EC_KEY *k,
+                            int param_format,
+                            int comp_format,
+                            int new_param_format,
+                            int new_comp_format,
+                            size_t key_src,
+                            size_t obj_type)
 {
-    size_t i;
-
-    for (i=0; i<OSSL_NELEM(known_results); ++i) {
-        if (strcmp(known_results[i].curve_name, known_curve_names[curve_i]) != 0)
-            continue;
-        if (param_formats_i[param_format_i] != known_results[i].param_format)
-            continue;
-        if (comp_formats_i[comp_format_i] != known_results[i].comp_format)
-            continue;
-        if (key_src != known_results[i].key_src)
-            continue;
-
-        break;
-    }
+    int rv = 0, res_asn1_flag, res_point_form, buf_len;
+    unsigned char *buf = NULL, *bufi;
+    EC_KEY *k2 = NULL;
+    EC_GROUP *g = NULL;
+    const EC_GROUP *gref = NULL;
+
+    /*
+     * Duplicate k so our changing parameters below does not
+     * affect other tests
+     */
+    k = EC_KEY_dup(k);
+    if (k == NULL)
+        goto fail;
 
-    if (i >= OSSL_NELEM(known_results)) {
-        fprintf(stderr, "# could not identify known result\n");
-        return 0;
+    if (new_param_format >= 0)
+        EC_KEY_set_asn1_flag(k, new_param_format);
+    else
+        new_param_format = param_format;
+
+    if (new_comp_format >= 0)
+        EC_KEY_set_conv_form(k, new_comp_format);
+    else if (key_src == KEY_SRC_GENERATE)
+        /*
+         * If we just generated our key we effectively already set the format,
+         * so we need to expect that rather than the default. Otherwise expect
+         * reversion to uncompressed.
+         */
+        new_comp_format = comp_format;
+    else
+        new_comp_format = POINT_CONVERSION_UNCOMPRESSED;
+
+    if (obj_type == OBJ_TYPE_KEY) {
+        /* If we are checking a private key: */
+        if (!EC_KEY_check_key(k))
+            goto fail;
+
+        buf_len = i2d_ECPrivateKey(k, &buf);
+        if (buf_len <= 0 || buf == NULL)
+            goto fail;
+
+        bufi = buf;
+        k2 = d2i_ECPrivateKey(NULL, (const unsigned char **)&bufi, buf_len);
+        if (k2 == NULL)
+            goto fail;
+
+        /*
+         * When we put keys through serialization and then deserialization, they
+         * always become uncompressed. (The same is not true for parameters,
+         * however.)
+         */
+        new_comp_format = POINT_CONVERSION_UNCOMPRESSED;
```

Comment:
This is surprising in 1.1.1: if we ask for compressed point serialization we should get the output in compressed format, and the decoding process should detect the input is in compressed form and retain that information.

@mattcaswell any idea what might be going on?